extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/impl/DbCheckingPMetaSupplier.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2023 Hazelcast Inc.
+ *
+ * Licensed under the Hazelcast Community License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hazelcast.jet.mongodb.impl;
+
+import com.hazelcast.cluster.Address;
+import com.hazelcast.dataconnection.DataConnection;
+import com.hazelcast.dataconnection.impl.InternalDataConnectionService;
+import com.hazelcast.function.SupplierEx;
+import com.hazelcast.jet.JetException;
+import com.hazelcast.jet.core.ProcessorMetaSupplier;
+import com.hazelcast.jet.core.ProcessorSupplier;
+import com.hazelcast.jet.core.Vertex;
+import com.hazelcast.jet.datamodel.Tuple2;
+import com.hazelcast.jet.impl.processor.ExpectNothingP;
+import com.hazelcast.jet.impl.util.Util;
+import com.hazelcast.jet.mongodb.dataconnection.MongoDataConnection;
+import com.hazelcast.jet.pipeline.DataConnectionRef;
+import com.hazelcast.spi.impl.NodeEngineImpl;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoDatabase;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.security.Permission;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.function.Function;
+
+import static com.hazelcast.internal.util.UuidUtil.newUnsecureUuidString;
+import static com.hazelcast.jet.datamodel.Tuple2.tuple2;
+import static com.hazelcast.jet.impl.util.Util.arrayIndexOf;
+import static com.hazelcast.jet.mongodb.impl.MongoUtilities.checkCollectionExists;
+import static com.hazelcast.jet.mongodb.impl.MongoUtilities.checkDatabaseExists;
+import static com.hazelcast.partition.strategy.StringPartitioningStrategy.getPartitionKey;
+import static java.util.Collections.singletonList;
+
+/**
+ * A {@link ProcessorMetaSupplier} that will check if requested database and collection exist before creating
+ * the processors.
+ */
+public class DbCheckingPMetaSupplier implements ProcessorMetaSupplier {
+
+    private final Permission requiredPermission;
+    private final boolean shouldCheck;
+    private boolean forceTotalParallelismOne;
+    private final String databaseName;
+    private final String collectionName;
+    private final ProcessorSupplier processorSupplier;
+    private final SupplierEx<? extends MongoClient> clientSupplier;
+    private final DataConnectionRef dataConnectionRef;
+    private int preferredLocalParallelism = Vertex.LOCAL_PARALLELISM_USE_DEFAULT;
+
+    private transient Address ownerAddress;
```

Comment:
this PMS is not reusable but we prefer reusable PMSes

Reply:
WDYM by "not reusable"? Yes, it's a specific use case within Mongo connector, same as e.g. ElasticSourcePMetaSupplier.

Reply:
I mean `ProcessorMetaSupplier.isReusable()`

Reply:
True, it's not reusable - however it's inevitable in this case IMHO. I think performance penalty from using non-reusable PMS is lower than performance gain (as I said, typical query is 2 checks instead of 2 * no of server nodes, I/O here was very big)